core/src/mast/node/basic_block_node/arbitrary.rs

```diff
@@ -194,42 +198,214 @@ impl Default for MastForestParams {
     }
 }
 
-/// Strategy for generating MastForest instances
-pub fn mast_forest_strategy(params: MastForestParams) -> impl Strategy<Value = MastForest> {
-    // BasicBlockNode generation must reference decorator IDs in [0, decorators)
-    let bb_params = BasicBlockNodeParams {
-        max_decorator_id_u32: params.decorators,
-        ..Default::default()
-    };
+impl Arbitrary for MastForest {
+    type Parameters = MastForestParams;
+    type Strategy = BoxedStrategy<Self>;
 
-    // 1) Generate a Vec<BasicBlockNode> with length in `params.blocks`
-    prop::collection::vec(any_with::<BasicBlockNode>(bb_params), params.blocks.clone())
-        // 2) Map concrete blocks -> build a concrete MastForest
-        .prop_map(move |blocks| {
-            let mut forest = MastForest::new();
+    fn arbitrary_with(params: Self::Parameters) -> Self::Strategy {
+        // BasicBlockNode generation must reference decorator IDs in [0, decorators)
+        let bb_params = BasicBlockNodeParams {
+            max_decorator_id_u32: params.decorators,
+            ..Default::default()
+        };
+
+        // 1) Generate a Vec<BasicBlockNode> with length in `params.blocks`
+        (
+            prop::collection::vec(any_with::<BasicBlockNode>(bb_params), params.blocks.clone()),
+            prop::collection::vec(any::<Decorator>(), params.decorators as usize..=params.decorators as usize)
+        )
+            // 2) Map concrete blocks -> build a concrete MastForest
+            .prop_map(move |(blocks, decorators)| {
+                let mut forest = MastForest::new();
+
+                // Pre-populate the decorator ID space so referenced IDs are valid.
+                // Generate all decorator types for more comprehensive testing
+                for decorator in decorators {
+                    forest
+                        .add_decorator(decorator)
+                        .expect("Failed to add decorator");
+                }
+
+                // Insert the generated blocks into the forest
+                for block in blocks {
+                    forest.add_node(block).expect("Failed to add block");
```

Comment:
Should we make each of those procedure roots (with `MastForest::make_root()`)? Both for the serde tests, as well as future proptests that rely on MAST forest procedures.